src/Microsoft.SqlTools.ServiceLayer/QueryExecution/SerializationService.cs

```diff
@@ -0,0 +1,256 @@
+                if (!hasSerializer)
+                {
+                    serializer = new DataSerializer(serializeParams);
+                    inProgressSerializations.AddOrUpdate(serializer.FilePath, serializer, (key, old) => serializer);
+                }
+                Func<Task<SerializeDataResult>> writeData = () =>
+                {
+                    return Task.Factory.StartNew(() =>
+                    {
+                        var result = serializer.ProcessRequest(serializeParams);
+                        if (serializeParams.IsComplete)
+                        {
+                            // Cleanup the serializer
+                            this.inProgressSerializations.TryRemove(serializer.FilePath, out serializer);
+                        }
+                        return result;
+                    });
+                };
+                await HandleRequest(writeData, requestContext, "HandleSerializeDataRequest");
+            }
+            catch (Exception ex)
+            {
+                await requestContext.SendError(ex.Message);
+            }
+        }
+
+        private async Task HandleRequest<T>(Func<Task<T>> handler, RequestContext<T> requestContext, string requestType)
+        {
+            Logger.Write(TraceEventType.Verbose, requestType);
+
+            try
+            {
+                T result = await handler();
+                await requestContext.SendResult(result);
+            }
+            catch (Exception ex)
+            {
+                await requestContext.SendError(ex.Message);
+            }
+        }
+    }
+    class DataSerializer
+    {
+        private IFileStreamWriter writer;
+        private SerializeDataRequestParams requestParams;
+        private IList<DbColumnWrapper> columns;
+
+        public string FilePath { get; private set; }
+
+        public DataSerializer(SerializeDataRequestParams requestParams)
+        {
+            this.requestParams = requestParams;
+            this.columns = this.MapColumns(requestParams.Columns);
+            this.FilePath = requestParams.FilePath;
+        }
+
+        private IList<DbColumnWrapper> MapColumns(ColumnInfo[] columns)
+        {
+            List<DbColumnWrapper> columnWrappers = new List<DbColumnWrapper>();
+            foreach (ColumnInfo column in columns)
+            {
+                DbColumnWrapper wrapper = new DbColumnWrapper(column);
+                columnWrappers.Add(wrapper);
+            }
+            return columnWrappers;
+        }
+
+
+        public SerializeDataResult ProcessRequest(SerializeDataRequestParams serializeParams)
+        {
+            SerializeDataResult result = new SerializeDataResult()
+            {
+            };
+            try
+            {
+                this.WriteData(serializeParams.Rows, serializeParams.IsComplete);
+                if (serializeParams.IsComplete)
+                {
+                    this.CloseStreams();
+                }
```

Comment:
This seems odd to do the cleanup here - wouldn't most things only call this request once? I would expect that after serialization is complete the streams would be automatically closed

Reply:
Ok so looking at it more it seems like this will get multiple requests in batches for a single file (which makes sense). I guess this is fine but I'm not completely sold on the IsComplete name - IsLast seemed to make more sense to me. But IsLast isn't great either - maybe IsLastBatch? I'd also put a comment about keeping the streams open until we get told that all the data is transferred since it's otherwise unclear that this is a batched operation. 

---
In reply to: [309756726](https://github.com/microsoft/sqltoolsservice/pull/840#discussion_r309756726) [](ancestors = 309756726)

Reply:
Ack, will consider renaming.